### ISSUE TITLE
Improve the compilation cache invalidation options

### ISSUE THREAD

**Issue by Cerdic** [scssphp/scssphp#268]
Linked to https://github.com/scssphp/scssphp/issues/256#issuecomment-724747705

There is now 2 usages of the Cache in ScssPhp : 
- in the Parser https://github.com/scssphp/scssphp/blob/master/src/Parser.php#L236 to get the result of an already parsed content
- in the Compiler https://github.com/scssphp/scssphp/blob/master/src/Compiler.php#L346 to get the result of an already Compiled content

For the Parser, the invalidation is straightforward: the result only rely on the content itself, and if content is unchanged it's safe to rely on the parsed result in cache.

For the Compiler, the invalidation is a bit more tricky. Even if the content of file is unchanged, the compilation result can change if one `@import`ed file is modified, which can happen in 2 cases:
1. the file that was imported in the previous compilation has changed, so it need to be re-parsed and then the whole parent file need to be compiled again
2. the `@import` resolving itself doesn't produce the same result (because the path changed, or because a new file is now available on the disk in a higher priority part of the path - ie overriding)

The first case is adressed by checking the timestamp of files that where used to produce the compiled cache: in the cache we are storing all imported files, and thus the Compiler can check if any of the files has been modified since the compilation result was stored in cache https://github.com/scssphp/scssphp/blob/master/src/Compiler.php#L350

The second case is not adressed at all at the moment.
A different path betwen the cached compilation result and the present call could be easy to manage (and I have to confess I didn't included that in ScssPHP because I was managing this issue in my calling function).

Checking that all `@import` would resolve the same way to detect the availability of a new file needs more informations to be stored in the cache:
- basically we need to register every call to `findImport()` with the `$url` argument and the `$this->currentDirectory` context and the result, store this list in the cache
- then when checking for the validity of a cached Compile result, go through this list, and check that all the `findImport()` would give the same final result
- if one of the `findImport()` gives a different filepath as a result OR if the timestamp changed on this file, then the cached Compile result should be ignored.

This process can be time consuming cause needing a lot of disk access, thus it should be an option for the Compiler to make a fast check (as now, only on the timestamp) or a deep check (checking all `findImport()`)

**Comment by mahagr** [comment]
Looks like we are on the same page (or you did read my mind). See my comment in the linked discussion. :)

**Comment by stof** [comment]
> For the Parser, the invalidation is straightforward: the result only rely on the content itself, and if content is unchanged it's safe to rely on the parsed result in cache.

I would add that we should take the library version as a cache busting too

**Comment by mahagr** [comment]
As you said, parser part is really easy. The compiler part was handled by Gantry itself but is currently broken as Gantry didn't use the cache before but had overrides to the `Compiler` class to partially (but not fully) resolve the import issue.

For the complete solution, I agree with the solution. A configuration option, which enables calls to `findImport()` for every single file.

For short term this may not be required, though. If there was an option to turn off Compiler cache, I would be able to use my custom logic to record imports myself and force Compiler to ignore cache by using another configuration option.

**Comment by Cerdic** [comment]
Also linked to #222 and #237 : `CompilationResult` class is definitly needed and could
- store the `findImport` stack in addition of the sourceMap
- have a `checkValid($fast=true)` method that could maybe be even not called by the Compiler itself but the caller with it's own preference about the `$fast` option?

**Comment by mahagr** [comment]
@Cerdic Regarding to your last comment, I have "`checkValid()`" method outside of the library with two custom ways to check the validity. One does not check any files, but looks if Gantry has been updated or if user changed configuration which contains the styling information for the theme.

**Comment by stof** [comment]
I'm wondering whether we should even have a caching layer for the compiled content inside the compiler itself. dart-sass does not have anything like that.

**Comment by mahagr** [comment]
I was wondering the same, right now I'm handling the compiler level caching myself. That said, it is quite hard to implement it right and the plan laid by @Cerdic looks really good and is harder to implement outside the Compiler...

**Comment by Cerdic** [comment]
Fixed by #270
